src/bitcoin.cpp

```diff
@@ -0,0 +1,204 @@
+        } else if (arg == "-m" || arg == "--multiprocess") {
+            cmd.use_multiprocess = true;
+        } else if (arg == "-M" || arg == "--monolithic") {
+            cmd.use_multiprocess = false;
+        } else if (arg == "-v" || arg == "--version") {
+            cmd.show_version = true;
+        } else if (arg == "-h" || arg == "--help" || arg == "help") {
+            cmd.show_help = true;
+        } else if (cmd.show_help && (arg == "-a" || arg == "--all")) {
+            cmd.show_help_all = true;
+        } else if (arg.starts_with("-")) {
+            throw std::runtime_error(strprintf("Unknown option: %s", arg));
+        } else if (!arg.empty()) {
+            cmd.command = arg;
+        }
+    }
+    return cmd;
+}
+
+//! Execute the specified bitcoind, bitcoin-qt or other command line in `args`
+//! using src, bin and libexec directory paths relative to this executable, where
+//! the path to this executable is specified in `wrapper_argv0`.
+//!
+//! @param args Command line arguments to execute, where first argument should
+//!             be a relative path to a bitcoind, bitcoin-qt or other executable
+//!             that will be located on the PATH or relative to wrapper_argv0.
+//!
+//! @param wrapper_argv0 String containing first command line argument passed to
+//!                      main() to run the current executable. This is used to
+//!                      help determine the path to the current executable and
+//!                      how to look for new executables.
+//
+//! @note This function doesn't currently print anything but can be debugged
+//! from the command line using strace or dtrace like:
+//!
+//!     strace -e trace=execve -s 10000 build/bin/bitcoin ...
+//!     dtrace -n 'proc:::exec-success  /pid == $target/ { trace(curpsinfo->pr_psargs); }' -c ...
+static void ExecCommand(const std::vector<const char*>& args, std::string_view wrapper_argv0)
+{
+    // Construct argument string for execvp
+    std::vector<const char*> exec_args{args};
+    exec_args.emplace_back(nullptr);
+
+    // Try to call ExecVp with given exe path.
+    auto try_exec = [&](fs::path exe_path, bool allow_notfound = true) {
+        std::string exe_path_str{fs::PathToString(exe_path)};
+        exec_args[0] = exe_path_str.c_str();
+        if (util::ExecVp(exec_args[0], (char*const*)exec_args.data()) == -1) {
+            if (allow_notfound && errno == ENOENT) return false;
+            throw std::system_error(errno, std::system_category(), strprintf("execvp failed to execute '%s'", exec_args[0]));
+        }
+        throw std::runtime_error("execvp returned unexpectedly");
+    };
+
+    // Get the wrapper executable path.
+    const fs::path wrapper_path{util::GetExePath(wrapper_argv0)};
+
+    // Try to resolve any symlinks and figure out the directory containing the wrapper executable.
+    std::error_code ec;
+    fs::path wrapper_dir{fs::weakly_canonical(wrapper_path, ec)};
```

Comment:
in commit nit: since the error code `ec` isn't used, could use the shorter variant of `weakly_canonical` (denoted as (3) in https://en.cppreference.com/w/cpp/filesystem/canonical)
```suggestion
    fs::path wrapper_dir{fs::weakly_canonical(wrapper_path)};
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31375#discussion_r2101326882

Thanks, again it's important to call the non-throwing overload here, because any error should be handled by trying to execute relative to the original, not throwing an error if symlinks can't be resolved for whatever reason.